src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java

```diff
@@ -0,0 +1,189 @@
+          getPoLineById(poLineId, lang, httpClient, ctx, okapiHeaders, logger).thenAccept(poLineJson -> {
+            PoLine poLine = poLineJson.mapTo(PoLine.class);
+            calculatePoLineReceiptStatus(poLine, listOfPieces)
+              .thenCompose(status -> updatePoLineReceiptStatus(poLine, status, httpClient, ctx, okapiHeaders, logger))
+              .thenAccept(future::complete);
+          })
+            .exceptionally(e -> {
+              logger.error("The error getting poLine by id {}", poLineId, e);
+              future.completeExceptionally(e);
+              return null;
+            });
+        })
+          .exceptionally(e -> {
+            logger.error("The error happened getting all pieces by poLine {}", poLineId, e);
+            future.completeExceptionally(e);
+            return null;
+          });
+      } else {
+        logger.debug("Receipt status of pieces and po line are consistent");
+        future.complete(null);
+      }
+    })
+      .exceptionally(e -> {
+        logger.error("The error happened getting a piece by id {}", pieceIdUpdate, e);
+        future.completeExceptionally(e);
+        return null;
+      });
+
+    // Now wait for all operations to be completed and send reply
+    completeAllFutures(ctx, futures, message);
+  }
+
+
+  private CompletableFuture<PoLine.ReceiptStatus> calculatePoLineReceiptStatus(PoLine poLine,
+      List<org.folio.rest.acq.model.Piece> pieces) {
+
+    if (pieces.isEmpty())
+      return completedFuture(poLine.getReceiptStatus());
+    else {
+      return getPiecesQuantityByPoLineAndStatus(ReceivingStatus.EXPECTED, pieces)
+        .thenCompose(expectedQty -> calculatePoLineReceiptStatus(expectedQty, poLine, pieces))
+        .exceptionally(e -> {
+          logger.error("The expected receipt status for PO Line '{}' cannot be calculated", e, poLine.getId());
+          return null;
+        });
+    }
+  }
+
+  private CompletableFuture<ReceiptStatus> calculatePoLineReceiptStatus(int expectedPiecesQuantity, PoLine poLine,
+      List<org.folio.rest.acq.model.Piece> pieces) {
+    if (!isCheckin(poLine) && expectedPiecesQuantity == 0) {
+      return CompletableFuture.completedFuture(FULLY_RECEIVED);
+    }
+    // Partially Received: In case there is at least one successfully received
+    // piece
+    if (StreamEx.of(pieces)
+      .anyMatch(piece -> ReceivingStatus.RECEIVED == piece.getReceivingStatus())) {
+      return CompletableFuture.completedFuture(PARTIALLY_RECEIVED);
+    }
+    // Pieces were rolled-back to Expected. In this case we have to check if
+    // there is any Received piece in the storage
+    return getPiecesQuantityByPoLineAndStatus(ReceivingStatus.RECEIVED, pieces)
+      .thenApply(receivedQty -> receivedQty == 0 ? AWAITING_RECEIPT : PARTIALLY_RECEIVED);
+  }
+
+  private CompletableFuture<Integer> getPiecesQuantityByPoLineAndStatus(ReceivingStatus receivingStatus, List<Piece> pieces) {
```

Comment:
Again, it's an enum, no need to compare as String...  Can be simplified:
```
    return CompletableFuture.supplyAsync(() -> (int) pieces.stream()
      .filter(piece -> piece.getReceivingStatus() == receivingStatus)
      .count());
```

Reply:
Done